lib/iris/tests/unit/data_manager/test_DataManager.py

```diff
@@ -0,0 +1,602 @@
+
+    def test_lazy_with_real__realised_dtype_clearance(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        mask_array = ma.arange(size).reshape(shape)
+        dtype = mask_array.dtype
+        lazy_array = as_lazy_data(mask_array)
+        dm = DataManager(lazy_array, realised_dtype=dtype)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        self.assertEqual(dm._realised_dtype, dtype)
+        self.assertEqual(dm.dtype, dtype)
+        dm.replace(mask_array)
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertIsNone(dm._realised_dtype)
+        self.assertEqual(dm.dtype, dtype)
+        self.assertIs(dm._real_array, mask_array)
+        self.assertIs(dm.data, mask_array)
+
+    def test_real_with_lazy__realised_dtype_setter(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        real_array = np.arange(size).reshape(shape)
+        mask_array = ma.masked_array(real_array) * 10
+        dtype = mask_array.dtype
+        lazy_array = as_lazy_data(mask_array)
+        dm = DataManager(real_array)
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertIsNone(dm._realised_dtype)
+        self.assertEqual(dm.dtype, dtype)
+        dm.replace(lazy_array, realised_dtype=dtype)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        self.assertEqual(dm._realised_dtype, dtype)
+        self.assertEqual(dm.dtype, dtype)
+        self.assertIs(dm._lazy_array, lazy_array)
+        self.assertArrayEqual(dm.data, mask_array)
+
+    def test_real_with_real__realised_dtype_failure(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        real_array = np.arange(size).reshape(shape)
+        dm = DataManager(real_array)
+        emsg = 'Cannot set realised dtype, no lazy data is available.'
+        with self.assertRaisesRegexp(ValueError, emsg):
+            dm.replace(real_array * 10, realised_dtype=real_array.dtype)
+        self.assertIs(dm._real_array, real_array)
+        self.assertArrayEqual(dm.data, real_array)
+
+
+class Test___copy__(tests.IrisTest):
+    def test(self):
+        dm = DataManager(np.array(0))
+        emsg = 'Shallow-copy of {!r} is not permitted.'
+        name = type(dm).__name__
+        with self.assertRaisesRegexp(copy.Error, emsg.format(name)):
+            copy.copy(dm)
+
+
+class Test___deepcopy__(tests.IrisTest):
+    def test(self):
+        dm = DataManager(np.array(0))
+        method = 'iris._data_manager.DataManager._deepcopy'
+        return_value = mock.sentinel.return_value
+        with mock.patch(method) as mocker:
+            mocker.return_value = return_value
+            result = copy.deepcopy(dm)
+            self.assertEqual(mocker.call_count, 1)
+        self.assertIs(result, return_value)
```

Comment:
Interesting that we don't have tests for cube._deepcopy 
Yep, I'm happy for this to be left for someother time. We don't gain much from doing it and I had suspected it would be quite painful. So in the name of diminishing returns I would agree to putting this off.